server/src/main/java/org/opensearch/index/translog/transfer/TranslogCkpFilesTransferManager.java

```diff
@@ -0,0 +1,116 @@
+import org.opensearch.core.index.shard.ShardId;
+import org.opensearch.index.remote.RemoteTranslogTransferTracker;
+import org.opensearch.index.translog.transfer.FileSnapshot.CheckpointFileSnapshot;
+import org.opensearch.index.translog.transfer.FileSnapshot.TransferFileSnapshot;
+import org.opensearch.index.translog.transfer.FileSnapshot.TranslogFileSnapshot;
+import org.opensearch.indices.RemoteStoreSettings;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Translog transfer manager to transfer {@link TranslogCheckpointSnapshot} by transfering {@link TranslogFileSnapshot} and {@link CheckpointFileSnapshot} separately
+ *
+ * @opensearch.internal
+ */
+public class TranslogCkpFilesTransferManager extends TranslogTransferManager {
+
+    TransferService transferService;
+    FileTransferTracker fileTransferTracker;
+
+    public TranslogCkpFilesTransferManager(
+        ShardId shardId,
+        TransferService transferService,
+        BlobPath remoteDataTransferPath,
+        BlobPath remoteMetadataTransferPath,
+        FileTransferTracker fileTransferTracker,
+        RemoteTranslogTransferTracker remoteTranslogTransferTracker,
+        RemoteStoreSettings remoteStoreSettings
+    ) {
+        super(
+            shardId,
+            transferService,
+            remoteDataTransferPath,
+            remoteMetadataTransferPath,
+            fileTransferTracker,
+            remoteTranslogTransferTracker,
+            remoteStoreSettings
+        );
+        this.transferService = transferService;
+        this.fileTransferTracker = fileTransferTracker;
+    }
+
+    @Override
+    public void transferTranslogCheckpointSnapshot(
+        Set<TranslogCheckpointSnapshot> toUpload,
+        Map<Long, BlobPath> blobPathMap,
+        LatchedActionListener<TranslogCheckpointSnapshot> latchedActionListener
+    ) throws Exception {
+        for (TranslogCheckpointSnapshot tlogAndCkpTransferFileSnapshot : toUpload) {
+            Set<TransferFileSnapshot> filesToUpload = new HashSet<>();
+            Set<Exception> exceptionList = ConcurrentCollections.newConcurrentSet();
+
+            String tlogFileName = tlogAndCkpTransferFileSnapshot.getTranslogFileName();
+            String ckpFileName = tlogAndCkpTransferFileSnapshot.getCheckpointFileName();
+
+            if (!fileTransferTracker.uploaded(tlogFileName)) {
+                filesToUpload.add(tlogAndCkpTransferFileSnapshot.getTranslogFileSnapshot());
+            }
+            if (!fileTransferTracker.uploaded(ckpFileName)) {
+                filesToUpload.add(tlogAndCkpTransferFileSnapshot.getCheckpointFileSnapshot());
+            }
+            assert !filesToUpload.isEmpty();
+            AtomicBoolean listenerProcessed = new AtomicBoolean();
+            final CountDownLatch latch = new CountDownLatch(filesToUpload.size());
```

Comment:
How is this latch different than the one in `LatchedActionListener` that is passed to this method as a parameter?

Reply:
Now in this method...We are processing one `generation's` files upload at a time i.e. we will transfer tlog and ckp file for a generation. While the current generation's files are uploading we don't want to block other generations we want this process to be async. Hence this latch is there to listen the transfer of files ckp and tlog. On the other hand the latch above given in the method parameter listen for a `generation` transfer event.